cpp/csp/adapters/websocket/ClientAdapterManager.cpp

```diff
@@ -24,57 +19,62 @@ ClientAdapterManager::ClientAdapterManager( Engine* engine, const Dictionary & p
 : AdapterManager( engine ), 
     m_active( false ), 
     m_shouldRun( false ), 
-    m_endpoint( nullptr ),
+    m_endpoint( std::make_unique<WebsocketEndpoint>( properties ) ),
     m_inputAdapter( nullptr ), 
     m_outputAdapter( nullptr ),
     m_updateAdapter( nullptr ),
     m_thread( nullptr ), 
     m_properties( properties ) 
-{
-    if( m_properties.get<bool>( "use_tls" ) ) 
-    {
-        m_endpoint = new WebsocketEndpointTLS( properties );
-    } 
-    else 
-    {
-        m_endpoint = new WebsocketEndpointNoTLS( properties );
-    }
-
-
-};
+{ };
 
 ClientAdapterManager::~ClientAdapterManager()
 { };
 
 void ClientAdapterManager::start( DateTime starttime, DateTime endtime )
 {
-    if(m_inputAdapter != nullptr)
-    {
-        m_endpoint -> setOnMessageCb( [ this ]( std::string msg ) {
-            PushBatch batch( m_engine -> rootEngine() );
-            m_inputAdapter -> processMessage( msg, &batch );
-        });
-    }
-    m_endpoint -> setOnOpenCb( [ this ]() {
-        m_active = true;
-        pushStatus( StatusLevel::INFO, ClientStatusType::ACTIVE, "Connected successfully" );
-    });
-    m_endpoint -> setOnFailCb( [ this ]() {
-        m_active = false;
-        pushStatus( StatusLevel::ERROR, ClientStatusType::CONNECTION_FAILED, "Connection failed, will try to reconnect" );
-    });
-    m_endpoint -> setOnCloseCb( [ this ]() {
-        m_active = false;
-        pushStatus( StatusLevel::INFO, ClientStatusType::CLOSED, "Connection closed" );
-    });
-    m_endpoint -> setOnSendFailCb( [ this ]( const std::string& s ) {
-        std::stringstream ss;
-        ss << "Failed to send: " << s;
-        pushStatus( StatusLevel::ERROR, ClientStatusType::MESSAGE_SEND_FAIL, ss.str() );
-    });
     AdapterManager::start( starttime, endtime );
-    // start the bg thread
+
     m_shouldRun = true;
+    m_endpoint -> setOnOpen(
+        [ this ]() {
+            m_active = true;
+            pushStatus( StatusLevel::INFO, ClientStatusType::ACTIVE, "Connected successfully" );
+        }
+    );
+    m_endpoint -> setOnFail(
+        [ this ]( const std::string& reason ) {
+            std::stringstream ss;
+            ss << "Connection Failure: " << reason;
+            m_active = false;
+            pushStatus( StatusLevel::ERROR, ClientStatusType::CONNECTION_FAILED, ss.str() );
+        } 
+    );
+    if( m_inputAdapter ) {
+        m_endpoint -> setOnMessage(
+            [ this ]( const std::string& msg ) {
+                PushBatch batch( m_engine -> rootEngine() );
+                m_inputAdapter -> processMessage( msg, &batch );
```

Comment:
this has to follow all the way through to avoid the copy
looks like you would need to pass void * and size_t len, and then no need for an extra copy at all